toolchain/base/value_store.h

```diff
@@ -79,24 +79,43 @@ constexpr StringId StringId::Invalid(StringId::InvalidIndex);
 
 // A simple wrapper for accumulating values, providing IDs to later retrieve the
 // value. This does not do deduplication.
-template <typename IdT>
+template <typename IdT, typename ValueT = typename IdT::IndexedType>
 class ValueStore {
  public:
   // Stores the value and returns an ID to reference it.
-  auto Add(typename IdT::IndexedType value) -> IdT {
+  auto Add(ValueT value) -> IdT {
     auto id = IdT(values_.size());
     values_.push_back(std::move(value));
     return id;
   }
 
+  // Adds a default constructed value and returns an ID to reference it.
+  auto AddDefaultValue() -> IdT {
+    auto id = IdT(values_.size());
+    values_.resize(id.index + 1);
+    return id;
+  }
+
+  // Returns a mutable value for an ID.
+  auto Get(IdT id) -> ValueT& {
+    CARBON_CHECK(id.index >= 0) << id.index;
+    return values_[id.index];
+  }
```

Comment:
I think we should distinguish between a `ValueStore` and a `MutableValueStore`. When storing (eg) `Integer`s, I think the guarantee of immutability is important, and we shouldn't provide any mutable access.

We could do this either with two different `ValueStore` types or by making the immutable ones have a `const`-qualified `ValueT` type. I think it's probably clearer to readers of the context / file APIs to have distinct type names for the `ValueStore`s, but I'd be happy either way.

Reply:
Having looked at the custom value stores in SemIR, many of which provide some kind of mutable access, I'm no longer so sure that putting the `const`ness in the name of the store will help readers very much. Just changing the `IndexedType`s for some of the `Id`s to be `const` seems fine to me.

Reply:
Done with `const`. I think it's mainly the shared ValueStores that can use this; the semir types have become more mutable.